File: solutions/020-valid-parentheses/solution.go
package main

import "fmt"

var pairs = map[byte]byte{
	')': '(',
	']': '[',
	'}': '{',
}

func isValid(s string) bool {
	stack := make([]byte, 0, len(s))

	for i := 0; i < len(s); i++ {
		ch := s[i]
		if open, ok := pairs[ch]; ok {
			if len(stack) == 0 || stack[len(stack)-1] != open {
				return false
			}
			stack = stack[:len(stack)-1]
		} else {
			stack = append(stack, ch)
		}
	}
	return len(stack) == 0
}

func main() {
	tests := []struct {
		s        string
		expected bool
	}{
		{"()", true},
		{"()[]{}", true},
		{"(]", false},
		{"([])", true},
		{"([)]", false},
		{"", true},
		{"(", false},
	}

	for _, tc := range tests {
		result := isValid(tc.s)
		status := "PASS"
		if result != tc.expected {
			status = "FAIL"
		}
		fmt.Printf("%s | s=%q => %v (expected %v)\n", status, tc.s, result, tc.expected)
	}
}

File: solutions/020-valid-parentheses/solution.py
class Solution:
    def is_valid(self, s: str) -> bool:
        pairs = {")": "(", "]": "[", "}": "{"}
        stack: list[str] = []

        for ch in s:
            if ch in pairs:
                if not stack or stack[-1] != pairs[ch]:
                    return False
                stack.pop()
            else:
                stack.append(ch)

        return not stack


if __name__ == "__main__":
    tests = [
        ("()", True),
        ("()[]{}", True),
        ("(]", False),
        ("([])", True),
        ("([)]", False),
        ("", True),
        ("(", False),
    ]
    sol = Solution()
    for s, expected in tests:
        result = sol.is_valid(s)
        status = "PASS" if result == expected else "FAIL"
        print(f"{status} | s={s!r} => {result} (expected {expected})")
